src/backend/distributed/commands/role.c

```diff
@@ -354,16 +354,18 @@ ExtractEncryptedPassword(Oid roleOid)
 
 	Datum passwordDatum = heap_getattr(tuple, Anum_pg_authid_rolpassword,
 									   pgAuthIdDescription, &isNull);
-	char *passwordCstring = TextDatumGetCString(passwordDatum);
 
-	table_close(pgAuthId, AccessShareLock);
-	ReleaseSysCache(tuple);
+	/* empty password is treated the same as NULL (no password) */
 
-	if (isNull)
+	const char *passwordCstring = "";
+	if (!isNull)
 	{
-		return NULL;
+		passwordCstring = TextDatumGetCString(passwordDatum);
```

Comment:
why are we 'changing' the behaviour from the previous intent?
we planned to return `NULL` when the password was set not set, eg `isNull` became true.

However, with this change we choose to return an empty string.

I do see that an empty string is treated as `NULL` password, which could explain how this works.
However, our tests seem to disagree with this statement:

> ALTER ROLE new_role PASSWORD '';
> NOTICE:  empty string is not a valid password, clearing password

Would it make sense to default to `NULL` here if `isNull` is set. We can easily move the pstrdup inside the scope of the branch here, and change the initialization of `passwordCstring` to `NULL`.

If that is not possible or desirable we need to have better documentation of its subtleties.

Reply:
> The reason why I 'changed' the behavior was because it's better to send the same command to the worker nodes. If we return NULL instead of an empty string, the workers would get this command ALTER ROLE new_role PASSWORD NULL, whereas in the coordinator we run ALTER ROLE new_role PASSWORD. Right now these two commands do the same thing, but maybe it will change in the future. I prefer to maintain consistency here.

Okay this is not true, in this way the other way around doesn't work. So the question is whether I should pick '' or NULL to send to the workers. You're right, considering the notice, it's better to default to NULL.

Reply:
For future reference: This is because of this fix for a security issue I reported ~6 years ago to Postgres: https://github.com/postgres/postgres/commit/bf6b9e94445610a3d84cf9521032fab993f96fd6